File: d02/main.py
#!/usr/bin/env python3

from sys import argv
from time import process_time

tasks: list[int] = [1, 2]
test_only: bool = True

def split_vec(vec: list[int]) -> list[list[int]]:
    ret = []
    for i in range(len(vec)):
        ret.append(vec[0:i] + vec[i+1:])
    return ret

def is_safe(report: list[int], ident="") -> (bool, int): # type: ignore
    # print(f"{ident}checking report part {report}", end="")
    if (len(report) == 1):
        # print(" -> True")
        return True, 1
    sign = report[1] - report[0]
    for i in range(len(report)-1):
        diff = report[i+1] - report[i]
        if diff == 0 or diff*sign < 0 or abs(diff) > 3:
            # print(" -> False: ", i)
            return False, i
    # print(" -> True")
    return True, 1

def solve():
    inputf = "input_testdata.txt" if test_only else "input.txt"

    start = process_time()

    puzzle = open(inputf).read()
    lines = puzzle.strip().splitlines()

    load = process_time()

    # action
    if 1 in tasks:
        safe_cnt = 0
        for l in lines:
            vec = [int(n) for n in l.split()]
            safe, idx = is_safe(vec)
            if safe:
                safe_cnt +=1
        print(f"Task 1: Out of {len(lines)} reports, {safe_cnt} are safe")
    
    if 2 in tasks:
        safe_cnt = 0
        safe_lines = []
        # safe_lines2 = []
        for num, l in enumerate(lines):
            vec = [int(n) for n in l.split()]
            safe, idx = is_safe(vec)
            if safe: 
                safe_lines.append(num)
                # safe_lines2.append(num)
                safe_cnt +=1
            else:

                # brute force, correct
                vv = split_vec(vec)
                for i, v in enumerate(vv):
                    ident = f"  {i}: "
                    safe, ign = is_safe(v, ident)
                    if safe:
                        safe_lines.append(num)
                        safe_cnt +=1
                        break
                
                # smart, but wrong
                # vec1 = vec[0:idx] + vec [idx+1:]
                # vec2 = vec[0:idx+1] + vec [idx+2:]
                # safe, idx = is_safe(vec1, " 1:  ")
                # if safe:
                #     safe_lines2.append(num)
                # else:
                #     safe, idx = is_safe(vec2, " 2:  ")
                #     if safe:
                #         safe_lines2.append(num)

    print(f"Taks 2: Out of {len(lines)} reports, {safe_cnt} are safe")
    # print(len(safe_lines))
    # print(len(safe_lines2))
    # print(list(set(safe_lines) - set(safe_lines2)))

    stop = process_time()

    print(f"Load took {load-start:3.6f} secs, calculation {stop-load:3.6f} secs")


def main(argv):
    global tasks, test_only

    if "-h" in argv or "--help" in argv:
        print("usage: python3 main.py [-h] [-s] [-1|2]")
        print("  -h:    this help")
        print("  -s:    execute large scale problem, not just example/test set, which is the default")
        print("  -1|2:  task 1 or 2, respectively (default: both tasks)")
        exit(1)

    tasks = [2] if "-2" in argv else [1] if "-1" in argv else [1,2]

    test_only = False if "-s" in argv else True

    print(f"\n===> Running task(s) {tasks} in {'test' if test_only else 'full'} mode ===================================\n")

    solve()


if __name__ == "__main__":
    main(argv)

    # read, eg, 5,6, and return as list of int tupels
    # input = "5,6\n3,4\n2,3"
    # lst = list(tuple((int(x), int(y))) for x, y in re.findall('([0-9]+),([0-9]+)', input))

    # read alphaanumeric (incl '_'), and return as list of lists
    # input = "key -> value\nk2 -> val_2\nk3 -> v3"
    # lst = list( [ x, y ] for x, y in re.findall('([\w]+) -> ([\w]+)', input))
    # print (lst)

    # ... or as dict
    # dct = dict( [ x, y ] for x, y in re.findall('([\w]+) -> ([\w]+)', input))
    # print(dct)

File: d02/main.js
#!/usr/bin/env node

// Note: I am just learning JavaScript (not from "ground up", I know several languages like C, C++, Java, Python, ...),
// but more like how to work best with lists, dictionaires, etc., and getting "hands on experience"
// You will find a lot of boilerplate code here, and also some of my solutions might suffer from the fact that
// I am still learning. You have been warned... ;-)

import { readFileSync } from 'node:fs';
import { exit } from 'node:process';

function load_data(test_only) {
    const dir = import.meta.dirname
    if (test_only) {
        return readFileSync(dir + '/input_testdata.txt', 'utf8')
    } else {
        return readFileSync(dir + '/input.txt', 'utf8')
    }
}

if (process.argv.includes("-h") || process.argv.includes("--help")) {
    console.log("usage: node main.js [-h] [-s] [-1|2]")
    console.log("  -h:    this help")
    console.log("  -s:    execute large scale problem, not just example/test set, whiich is the default")
    console.log("  -1|2:  task 1 or 2 only, repectively (default: both tasks)")
    exit(1)
}

let tasks = process.argv.includes("-2") ? [2] : process.argv.includes("-1") ? [1] : [1, 2]
let test_only = process.argv.includes("-s") ? false : true

let start = performance.now()

// As part of my "boilerplate" I read data into an array of lines of strings
// and keep track of the dimension (m = number of columns, n = number of rows)

let data = load_data(test_only)
let lines = data.trim().split('\n')
let n = lines.length
let m = lines[0].length
console.log(`Size: ${n}x${m}`)

let load = performance.now()

// Actual problem solving starts here ===========================================================================

function is_safe(report) {
    if (report.length > 1) {
        let sign = report[1] - report[0]
        for (let i = 0; i < report.length-1; i++) {
            const diff = report[i+1] - report[i]
            if (diff == 0 || diff * sign < 0 || Math.abs(diff) > 3) {
                // either 0 or change of sign or diff too big => error
                return { valid: false, errorIndex: i }
            }
        }
    }
    return { valid: true, errorIndex: 0 }
}

// Part 1 is straightforward: Iterate over all lines and check if the
// corresponding report/vector is safe. Store unsafe items as well as the 
// index where the error occured - this will be reused for part 2

let unsafe_lines = {}
let safe_cnt = 0

if (tasks.includes(1)) {
    for (let i = 0; i < lines.length; i++) {
        let vec = lines[i].split(" ").map(Number)
        let res = is_safe(vec)
        if (!res.valid) {
            unsafe_lines[i] = res
        } 
        // console.log(`checking report [${vec}] -> ${res.valid}: ${res.errorIndex}`)
    }
    safe_cnt = lines.length - Object.keys(unsafe_lines).length
    console.log(`Part 1: Out of ${lines.length} reports, ${safe_cnt} are safe`)    
}

// Part 2 reuses results of part 1: For all lines that have been identified
// as unsafe, try if they can be fixed by taking out respective elements
// Note: Part 1 and 2 can easily be combined to make this more efficient

if (tasks.includes(2)) {
    if (!tasks.includes(1)) {
        console.log("run task 1 first, as we need its results")
        exit(1)
    }
    for (let idx of Object.keys(unsafe_lines)) {
        let errVec = lines[idx].split(" ").map(Number)
        let errIdx = unsafe_lines[idx].errorIndex

        // check if the error can be fixed by
        //  - taking out the element at errIdx
        //  - taking out the element at errIdx + 1
        //  - taking out the element 0 if errIdx == 1
        let vec = errVec.slice(0, errIdx).concat(errVec.slice(errIdx+1))
        let res = is_safe(vec)
        if (!res.valid) {
            vec = errVec.slice(0, errIdx+1).concat(errVec.slice(errIdx+2))
            res = is_safe(vec)
            if (!res.valid && errIdx == 1) {
                vec = errVec.slice(1)
                res = is_safe(vec)
            }
        }
        safe_cnt += res.valid ? 1 : 0 
    }
    console.log(`Part 2: Out of ${lines.length} reports, ${safe_cnt} are now safe`)    
}

let stop = performance.now()

console.log(`Load took ${((load-start)/1000).toFixed(6)} msecs, calculation ${((stop-load)/1000).toFixed(6)} msecs`)
